src/main/msc/emfat_file.c

```diff
@@ -291,44 +294,102 @@ static void emfat_add_log(emfat_entry_t *entry, int number, uint32_t offset, uin
     entry->curr_size = size;
     entry->max_size = entry->curr_size;
     entry->readcb = bblog_read_proc;
-    emfat_set_entry_cma(entry);
+    // Set file modification/access times to be the same, either the default date or that from the RTC
+    entry->cma_time[1] = cmaTime;
+    entry->cma_time[2] = cmaTime;
 }
 
 static int emfat_find_log(emfat_entry_t *entry, int maxCount)
 {
     int limit = flashfsIdentifyStartOfFreeSpace();
     int lastOffset = 0;
     int currOffset = 0;
+    int buffOffset;
+    int hdrOffset;
     int fileNumber = 0;
-    uint8_t buffer[18];
+    uint8_t buffer[HDR_BUF_SIZE];
     int logCount = 0;
+    char *timeHeader = "H Log start datetime:";
+    int timeHeaderMatched = 0;
+    int lenTimeHeader = strlen(timeHeader);
 
     for ( ; currOffset < limit ; currOffset += 2048) { // XXX 2048 = FREE_BLOCK_SIZE in io/flashfs.c
 
-        flashfsReadAbs(currOffset, buffer, 18);
+        flashfsReadAbs(currOffset, buffer, HDR_BUF_SIZE);
 
         if (strncmp((char *)buffer, "H Product:Blackbox", 18)) {
             continue;
         }
 
+        // The length of the previous record is now known
         if (lastOffset != currOffset) {
-            emfat_add_log(entry, fileNumber, lastOffset, currOffset - lastOffset);
+            // Record the previous entry
+            emfat_add_log(entry++, fileNumber++, lastOffset, currOffset - lastOffset);
+
+            logCount++;
+        }
+
+        // Find the "Log start datetime" entry, example encoding "H Log start datetime:2019-08-15T13:18:22.199+00:00"
+
+        buffOffset = 18;
+        hdrOffset = currOffset;
+
+        while (true) {
+            if (buffer[buffOffset++] == timeHeader[timeHeaderMatched]) {
+                // This matches the header we're looking for so far
+                if (++timeHeaderMatched == lenTimeHeader) {
+                    // Complete match so read date/time into buffer
+                    flashfsReadAbs(hdrOffset + buffOffset, buffer, HDR_BUF_SIZE);
+
+                    // Extract the time values to create the CMA time
+                    char *nextToken = (char *)buffer;
+                    int year = strtoul(nextToken, &nextToken, 10);
+                    int month = strtoul(++nextToken, &nextToken, 10);
+                    int day = strtoul(++nextToken, &nextToken, 10);
+                    int hour = strtoul(++nextToken, &nextToken, 10);
+                    int min = strtoul(++nextToken, &nextToken, 10);
+                    int sec = strtoul(++nextToken, NULL, 10);
+
+                    // Set the file creation time
+                    if (year) {
+                        entry->cma_time[0] = EMFAT_ENCODE_CMA_TIME(day, month, year, hour, min, sec);
```

Comment:
We should also set the access and modification dates, not just the creation date.

Reply:
OK, modification and access timestamps now set to match creation.